orangecontrib/text/widgets/owwordlist.py

```diff
@@ -0,0 +1,605 @@
+# pylint: disable=missing-docstring,no-name-in-module
+from contextlib import contextmanager
+from typing import Optional, List, Dict, Set
+import os
+
+import numpy as np
+
+from AnyQt.QtCore import Qt, QModelIndex, QItemSelectionModel, \
+    QItemSelection, QItemSelectionRange, Signal
+from AnyQt.QtGui import QKeySequence, QPalette, QColor, QPainter
+from AnyQt.QtWidgets import QListView, QSizePolicy, QGridLayout, QLineEdit, \
+    QRadioButton, QGroupBox, QToolButton, QMenu, QAction, QFileDialog, \
+    QStyledItemDelegate, QStyleOptionViewItem, QWidget
+
+from orangewidget.utils.listview import ListViewSearch
+
+from Orange.data import Table, StringVariable, Domain
+from Orange.widgets import gui
+from Orange.widgets.settings import Setting, DomainContextHandler, \
+    ContextSetting
+from Orange.widgets.utils.annotated_data import create_annotated_table
+from Orange.widgets.utils.itemmodels import DomainModel, ModelActionsWidget, \
+    PyListModel
+from Orange.widgets.widget import Input, Output, OWWidget, Msg
+
+
+@contextmanager
+def disconnected(signal, slot, connection_type=Qt.AutoConnection):
+    signal.disconnect(slot)
+    try:
+        yield
+    finally:
+        signal.connect(slot, connection_type)
+
+
+class WordList:
+    NotModified, Modified = range(2)
+
+    def __init__(self, name: str, words: List, filename: Optional[str] = None):
+        self.name = name
+        self.filename = filename
+        # library words
+        self.words = list(words)
+        # current words
+        self.cached_words = list(words)
+        # update rule changes current words
+        self.update_rule_flag = WordList.NotModified
+
+    @property
+    def flags(self) -> int:
+        # 0 - NotModified, 1 - Modified
+        return int(self.words != self.cached_words or
+                   self.update_rule_flag == WordList.Modified)
+
+    def as_dict(self) -> Dict:
+        return {"name": self.name,
+                "words": list(self.words),
+                "filename": self.filename}
+
+    @classmethod
+    def from_dict(cls, state: Dict) -> "WordList":
+        return WordList(state["name"],
+                        state["words"].copy(),
+                        filename=state.get("filename"))
+
+    @staticmethod
+    def generate_word_list_name(taken_names: List[str]) -> str:
+        default_name = "Untitled"
+        indices = {0}
+        for name in taken_names:
+            if name.startswith(default_name):
+                try:
+                    indices.add(int(name[len(default_name):]))
+                except ValueError:
+                    pass
+        index = min(set(range(max(indices) + 1 + 1)) - indices)
+        return f"{default_name} {index}"
+
+
+class UpdateRules:
+    INTERSECT, UNION, INPUT, LIBRARY = range(4)
+    ITEMS = ["Intersection", "Union", "Only Input", "Ignore Input"]
```

Comment:
I would have this as "Only input" and "Ignore input". I believe this is how we have it elsewhere. :)

Reply:
Fixed